### PR TITLE
build: make Train.ipynb collab ready

### PR DESCRIPTION
**Pull request by alexgarel**
Make Train.ipynb collab ready without disturbing local run.

### CODE REVIEW

File: lib/colab.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from pathlib import Path
+import re
+import subprocess
+import sys
+
+
+try:
```
Comment (streino) [inline lib/colab.py:7]: Do you need this since you have to do the same in the notebook anyway?

File: lib/colab.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  import google.colab
+  IN_COLAB = True
+except:
+  IN_COLAB = False
+
+
+# library not to install in collab
+COLLAB_EXCLUDE = {"tensorflow", "ipython", "notebook"}
```
Comment (streino) [inline lib/colab.py:15]: Were you able to install robotoff with `colab_pip_install`? I don't remember what exactly went wrong but I had a version conflict between one of robotoff's dependencies and libs pre-installed in collab.

File: experiments/Train.ipynb [hunk 3]
```diff
@@ -21,6 +21,35 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "try:\n",
+    "  import google.colab\n",
+    "  IN_COLAB = True\n",
+    "except:\n",
+    "  IN_COLAB = False\n",
+    "# eventual initialization for collab notebooks\n",
+    "assert IN_COLAB and not os.path.exists(\".git\"), \"Nothing to do\"\n",
```
Comment (streino) [inline experiments/Train.ipynb:37]: Will notebook continue to execute in "run all" (and similar) mode if assert raises here?

**Comment by streino** [comment]
Thanks for making the PR, I like how it's going to be easy to run stuff in Collab now :)

File: lib/colab.py (hunk 2 above)
Comment (alexgarel) [reply lib/colab.py:15]: No problem with robotoff on my side…

**Comment by alexgarel** [comment]
@streino I updated the PR it to an even less intrusive pattern.

Right now you will have to put `COLAB_BRANCH = "colab-ready"` to test it.

File: lib/colab.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        # init git repo and go in experiments
+        result = subprocess.run("git clone https://github.com/openfoodfacts/off-category-classification.git".split(), capture_output=True)
+        git_cloned = 1
+        if result.returncode != 0:
+            raise RuntimeError("Error on git clone", result.stderr.decode("utf-8"))
+        os.chdir(f"{initial}/off-category-classification/experiments")
+    if branch_name:
+        result = subprocess.run(f"git checkout {branch_name}".split(), capture_output=True) 
+        if result.returncode != 0:
+            # in
```
Comment (streino) [inline lib/colab.py:65]: Comment is a bit obscure :p

**Comment by streino** [comment]
How about adding a badge in the README to directly open a notebook in colab?
```
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/openfoodfacts/off-category-classification/blob/master/experiments/Train.ipynb)
```

File: lib/colab.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+def colab_pip_install():
+    """Install needed parts for collab notebooks"""
+    requirements = list(colab_requirements_iter())
+    return pip_install(requirements)
+
+
+def init_git(branch_name=None):
+    """Init git repo in collab"""
+    initial = os.getcwd()
+    if not initial.endswith("experiments") or not os.path.exists("../.git"):
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply lib/colab.py:52]: Even better ` git rev-parse --git-dir` return code.

I thus changed the code a bit to be less dependent on location.

**Pull request merged**